src/vm/coverage.rs

```diff
@@ -0,0 +1,224 @@
+use std::{
+    collections::{BTreeMap, HashMap, HashSet},
+    fs::File,
+    io::Write,
+};
+
+use serde_json::Value as JsonValue;
+use vm::types::QualifiedContractIdentifier;
+use vm::SymbolicExpression;
+
+use super::functions::define::DefineFunctionsParsed;
+
+pub struct CoverageReporter {
+    executed_lines: HashMap<QualifiedContractIdentifier, HashMap<u32, u64>>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct ContractFileInfo {
+    contract: String,
+    src_file: String,
+    executable_lines: Vec<u32>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct CoverageFileInfo {
+    coverage: HashMap<String, Vec<(u32, u64)>>,
+}
+
+impl CoverageReporter {
+    pub fn new() -> CoverageReporter {
+        CoverageReporter {
+            executed_lines: HashMap::new(),
+        }
+    }
+
+    #[cfg(not(feature = "developer-mode"))]
+    pub fn report_eval(
+        &mut self,
+        _expr: &SymbolicExpression,
+        _contract: &QualifiedContractIdentifier,
+    ) {
+    }
+
+    #[cfg(feature = "developer-mode")]
+    pub fn report_eval(
+        &mut self,
+        expr: &SymbolicExpression,
+        contract: &QualifiedContractIdentifier,
+    ) {
+        if expr.match_list().is_some() {
+            // don't count the whole list expression: wait until we've eval'ed the
+            //   list components
+            return;
+        }
+
+        // other sexps can only span 1 line
+        let line_executed = expr.span.start_line;
```

Comment:
I'm confused by this.  How would this be reported in the coverage data?

```
(+
   u1
   u2
   u3
 )
 ```
 
 Is this 5 lines, or one line?

Reply:
4 lines